### PR TITLE
added vmware metadata provider, closes #3507

### LINKED ISSUE #3507: Feature request: add vmware/cloud-init as a metadata provider
Support being able to read and store cloud-init metadata from VMWare's VMX https://github.com/vmware/cloud-init-vmware-guestinfo

### PR DESCRIPTION
**Pull request by brlbil**
cloud-init data from vmware guest info as it described in the link below
https://github.com/vmware/cloud-init-vmware-guestinfo

closes #3507

**- What I did**
Implemented a vmware metadata provider that leverages vmware guest info 
**- How I did it**
Implemented Provider interface. Added provider_vmware file and added vmware to cdrom providers in main.go 
**- How to verify it**
Guest info can be set for a vm using govc tool and can be tested in that vm as described.
https://github.com/vmware/cloud-init-vmware-guestinfo
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: pkg/metadata/main.go
Comment (rn) [inline pkg/metadata/main.go:82]: the CDROM provided should probably be the last in the list

File: pkg/metadata/provider_vmware.go
Comment (rn) [inline pkg/metadata/provider_vmware.go:40]: Please don't log in the Probe function. You may get a log messages if not running on VMWare

Comment (brlbil) [reply pkg/metadata/provider_vmware.go]: Sorry, it took so long to get back to this.
I fixed the things you suggested.

I changed the probed data as UserData in order to be in more sync with cdrom provider.
Also, I cleaned base64 decoding code a little.

Comment (brlbil) [reply pkg/metadata/provider_vmware.go]: @rn I believe the latest patches resolved the issues you pointed.

File: pkg/metadata/main.go
Comment (voima-eetu) [inline pkg/metadata/main.go:106]: I think cdrom should also be last in this switch-case. Otherwise if one has both 'vmware' and 'cdrom' as a provider it will use cdrom, not vmware provider.

Comment (brlbil) [reply pkg/metadata/main.go:104]: You are right cdrom being the last provider makes more sense, the latest commit fixes this

**Comment by brlbil** [comment]
I know it took a lot of time to fix missing bits and pieces. It sees all the checks are passing now, so can we merge this now?
@rn

**Comment by brlbil** [comment]
I know everyone is busy and it is not realistic to expect things always go fast for opensource projects.
But I would like to see this included in `k3os` asap. So can anyone take a look at the PR and get it rolling, please?

**Comment by voima-eetu** [comment]
> I know everyone is busy and it is not realistic to expect things always go fast for opensource projects.
> But I would like to see this included in `k3os` asap. So can anyone take a look at the PR and get it rolling, please?

A bit off-topic; I built k3os image with this your merge request and deployed it to VMware virtual datacenter using vcloud director, so thanks for this patch!

But for some reason my k3os installation on latest version of VMware cloud director wasn't able to read 'vmware' provider. On boot K3OS wasn't able to read guest-properties that I had added('guestinfo.user-data') and gave me just message "No metadata/userdata found. Bye".

Have you got this provider working? If yes, with what platform and which guest-properties?

**Comment by brlbil** [comment]
As you said I am sure we should discuss these issues here either. I think the problem is key should be set as `guestinfo.userdata` not `guestinfo.user-data`

**Comment by edvinerikson** [comment]
What's blocking this to be merged?

File: pkg/metadata/provider_vmware.go
Comment (edvinerikson) [inline pkg/metadata/provider_vmware.go:38]: Will this really be available if built with linuxkit? Wouldn't this tool only be available in the open-vm-tools image?

Comment (johnlaur) [reply pkg/metadata/provider_vmware.go:38]: It will be available if open-vm-tools is intalled, which is sort of a requirement if you want to make RPC calls through VMCI to get guestinfo data.... I dont understand what is being requested.

This mirrors the functionality of vmware's cloud-init https://github.com/vmware/cloud-init-vmware-guestinfo/blob/master/DataSourceVMwareGuestInfo.py which functions the same way.

Comment (brlbil) [reply pkg/metadata/provider_vmware.go:38]: Since edvinerikson has brought it up I had checked if it would be possible not to use the command-line tool and use a library instead. It is possible to use this [library](https://github.com/vmware/vmw-guestinfo). But this pull request has been idle for so long I am not sure maintainers have any interest in accepting it. I do not want to work on it if it is not going to be ignored.

Comment (robertkaelin) [reply pkg/metadata/provider_vmware.go:38]: I made a small PR to your source branch, as an example, with the said library 8 days ago: https://github.com/brlbil/linuxkit/pull/1
This would give you the possibility of removing the vmware-rpctool

**Comment by thehedgefrog** [comment]
Any updates to this?

**Comment by chris93111** [comment]
Any updates to this?

**Comment by observ3r** [comment]
Hello @rn - could you review/approve this?

**Comment by observ3r** [comment]
@deitch Could you take a look at this?

**Comment by deitch** [comment]
This looks pretty good, and you did address @rn's comments.

The only questions I have are:

1. Can you get @robertkaelin 's proposed changes [here](https://github.com/brlbil/linuxkit/pull/1) in, so that you do not have the dependency on `vmware-rpctool`? That would also mean documenting it, showing how to ensure it is available, etc. etc. His library solution is much better.
2. Is there any way to add a test for this? We only have a test for metadata on a CD [here](https://github.com/linuxkit/linuxkit/tree/master/test/cases/040_packages/013_metadata/000_cidata), so I can hardly be a stickler for it, but if it is possible, it would be good.

**Comment by deitch** [comment]
Completely separately, I see several people interested here. How are you using this (linuxkit in general)?

**Comment by brlbil** [comment]
I do not have access to a vcenter currently, so I will test the latest changes next week just to make sure everything works as expected.

**Comment by brlbil** [comment]
@deitch I have tested the package on a VMWare VM, and it works as expected.
Also, rebase and squash my commits as requested.

**Review by deitch — DISMISSED** [review]
LGTM

**Comment by deitch** [comment]
Looks fine to me. Let CI go green and we can merge in.

**Comment by KittyKatt** [comment]
@brlbil Any update on this? Really looking forward to configure k3os on vSphere without having to keep open a fork of linuxkit and build it every time I want a new ISO.

**Comment by brlbil** [comment]
@KittyKatt Unfortunately, I do not have any spare time to track the error that the pipeline throws. I have to do this on my own time. If anyone can help to track the error in the pipeline I would appreciate it.

**Comment by observ3r** [comment]
Appears the logs have expired for the Action that failed. Could it be re-run?

**Comment by observ3r** [comment]
@brlbil @deitch I don't have permission to rerun looks like. Could someone rerun it to generate the logs of the failed job?

**Comment by brlbil** [comment]
@observ3r I cannot run them either.

**Comment by KittyKatt** [comment]
> @KittyKatt Unfortunately, I do not have any spare time to track the error that the pipeline throws. I have to do this on my own time. If anyone can help to track the error in the pipeline I would appreciate it.

I tried looking into the errors when it was available and correcting/building locally, but unfortunately I am not anywhere near an expert in Go and couldn't make heads nor tails of it. Sorry!

**Comment by observ3r** [comment]
Hey @deitch , are you able to rerun it?

**Comment by deitch** [comment]
The rerun button isn't there. 🤷‍♂️ 

Might as well rebase it to get the parallelization we did for actions (runs much more quickly), and push it out, which will cause CI to kick off again anyways.

**Comment by observ3r** [comment]
Oh awesome, lookin good! I suppose just waiting on @rn at this point?

**Comment by observ3r** [comment]
@brlbil Wanna rebase it? I don't think I can

**Comment by brlbil** [comment]
@observ3r @deitch I rebased, but I think the workflow requires approval

**Review by deitch — APPROVED** [review]

**Comment by deitch** [comment]
Looks like something relating to the added packages failed? Maybe a `go mod tidy` followed by a `go mod vendor`? 🤷‍♂️

**Comment by brlbil** [comment]
I think I see what is going on. The `github.com/vmware/vmw-guestinfo` package does not support arm64 so it fails to build. We can open an issue upstream to add support for it. 

@deitch how to best avoid including VMware metadata in some of the architectures? I am pretty sure VMWare-tools is not supported on all the architectures that linuxkit is supporting like the arm.

**Comment by deitch** [comment]
Really? No arm64 support? That is surprising.

Do a dedicated file like `provider_vmware_amd64.go` and `provider_vmware_unsupported.go`, use `//build` directives at each, one for all other arch unsupported, one for amd64; for the amd64, return a valid struct, for all others, return `nil`, and add an `if == nil` check before adding it.

**Comment by brlbil** [comment]
Opened an issue for adding support for arm64 https://github.com/vmware/vmw-guestinfo/issues/25

@deitch I think we should proceed without arm64 support added. I have a feeling it would take a long time for it would be added upstream. If it is added then we can add support here as well. What do you think?

**Comment by deitch** [comment]
Agreed. Just do the `build` directives based on the suggestion above.

**Comment by brlbil** [comment]
@deitch ci passed on my fork, needs your touch here.

**Comment by brlbil** [comment]
I fixed if check, the provider was added to the list when it is nil.

**Review by deitch — APPROVED** [review]

**Pull request merged**
